dandiapi/api/services/embargo/__init__.py

```diff
@@ -33,28 +38,65 @@ def _delete_asset_blob_tags(client: S3Client, blob: str):
 
 # NOTE: In testing this took ~2 minutes for 100,000 files
 def _remove_dandiset_asset_blob_embargo_tags(dandiset: Dandiset):
-    # First we need to generate a CSV manifest containing all asset blobs that need to be untaged
-    embargoed_asset_blobs = AssetBlob.objects.filter(
-        embargoed=True, assets__versions__dandiset=dandiset
-    ).values_list('blob', flat=True)
+    embargoed_asset_blobs = list(
+        AssetBlob.objects.filter(embargoed=True, assets__versions__dandiset=dandiset).values_list(
+            'blob', flat=True
+        )
+    )
 
     client = get_boto_client(config=Config(max_pool_connections=100))
     with ThreadPoolExecutor(max_workers=100) as e:
-        for blob in embargoed_asset_blobs:
+        futures = [
             e.submit(_delete_asset_blob_tags, client=client, blob=blob)
+            for blob in embargoed_asset_blobs
+        ]
+
+    # Check if any failed and raise exception if so
+    failed = [
+        blob for i, blob in enumerate(embargoed_asset_blobs) if futures[i].exception() is not None
+    ]
+    if failed:
+        raise AssetTagRemovalError('Some blobs failed to remove tags', blobs=failed)
 
 
 @transaction.atomic()
-def _unembargo_dandiset(dandiset: Dandiset):
-    # NOTE: Before proceeding, all asset blobs must have their embargoed tags removed in s3
+def _unembargo_dandiset(ds: Dandiset):
+    logger.info('Unembargoing Dandiset %s', ds.identifier)
+    logger.info('\t%s assets', ds.draft_version.assets.count())
+
+    if ds.embargo_status != Dandiset.EmbargoStatus.UNEMBARGOING:
+        raise DandiError(
+            message=f'Expected dandiset state {Dandiset.EmbargoStatus.UNEMBARGOING}, found {ds.embargo_status}',  # noqa: E501
+            http_status_code=500,
+        )
+    if ds.uploads.all().exists():
+        raise DandisetActiveUploadsError(http_status_code=500)
+
+    # Remove tags in S3
+    logger.info('Removing tags...')
+    _remove_dandiset_asset_blob_embargo_tags(ds)
+
+    # Update embargoed flag on asset blobs
+    updated = AssetBlob.objects.filter(embargoed=True, assets__versions__dandiset=ds).update(
+        embargoed=False
+    )
+    logger.info('Updated %s asset blobs', updated)
+
+    # Set status to OPEN
+    Dandiset.objects.filter(pk=ds.pk).update(embargo_status=Dandiset.EmbargoStatus.OPEN)
+    logger.info('Dandiset embargo status updated')
```

Comment:
Do these operations need to be part of a transaction?

Reply:
This function (`unembargo_dandiset`) is wrapped in the `transaction.atomic()` decorator.